File: services/servico-order/src/application/OrderService.ts
import { Order, CreateOrderProps } from '../domain/Order';
import { OrderRepository } from './ports/OrderRepository';
import { EventPublisher } from './ports/EventPublisher';

export class OrderNotFoundError extends Error {
  constructor(id: string) {
    super(`Pedido "${id}" não encontrado.`);
    this.name = 'OrderNotFoundError';
  }
}

/**
 * Casos de uso do ServicoOrder (bounded context Order Management,
 * *core domain* do FTGO — ver EC-M2.1 e EC-M4.3).
 *
 * API síncrona exposta (EC-M4.3): POST /orders, GET /orders/{id}.
 * Eventos consumidos: CreditReserved, CreditRejected, TicketPrepared,
 * DeliveryScheduled (+ DeliveryCompleted, extensão documentada em
 * Order.ts). Nesta primeira versão, esses eventos chegam via um
 * EventBus em memória, simulando os demais microsserviços até que
 * sejam implementados.
 */
export class OrderService {
  constructor(
    private readonly repository: OrderRepository,
    private readonly eventPublisher: EventPublisher
  ) {}

  async createOrder(props: CreateOrderProps): Promise<Order> {
    const { order, event } = Order.create(props);
    await this.repository.save(order);
    await this.eventPublisher.publish(event);
    return order;
  }

  async getOrder(id: string): Promise<Order> {
    const order = await this.repository.findById(id);
    if (!order) {
      throw new OrderNotFoundError(id);
    }
    return order;
  }

  async handleCreditReserved(orderId: string, reservedAmountCents: number): Promise<Order> {
    const order = await this.getOrder(orderId);
    const event = order.onCreditReserved(reservedAmountCents);
    await this.repository.save(order);
    await this.eventPublisher.publish(event);
    return order;
  }

  async handleCreditRejected(orderId: string, reason: string): Promise<Order> {
    const order = await this.getOrder(orderId);
    const event = order.onCreditRejected(reason);
    await this.repository.save(order);
    await this.eventPublisher.publish(event);
    return order;
  }

  async cancelOrder(orderId: string, reason: string): Promise<Order> {
    const order = await this.getOrder(orderId);
    const event = order.cancel(reason);
    await this.repository.save(order);
    await this.eventPublisher.publish(event);
    return order;
  }

  async handleTicketPrepared(orderId: string): Promise<Order> {
    const order = await this.getOrder(orderId);
    order.onTicketPrepared();
    await this.repository.save(order);
    return order;
  }

  async handleDeliveryScheduled(orderId: string, courierId: string): Promise<Order> {
    const order = await this.getOrder(orderId);
    order.onDeliveryScheduled(courierId);
    await this.repository.save(order);
    return order;
  }

  async handleDeliveryCompleted(orderId: string): Promise<Order> {
    const order = await this.getOrder(orderId);
    const event = order.onDeliveryCompleted();
    await this.repository.save(order);
    await this.eventPublisher.publish(event);
    return order;
  }
}

File: services-java/servico-order/src/main/java/br/com/ftgo/order/application/OrderService.java
package br.com.ftgo.order.application;

import java.util.List;

import org.springframework.stereotype.Service;

import br.com.ftgo.order.application.port.EventPublisher;
import br.com.ftgo.order.application.port.OrderRepository;
import br.com.ftgo.order.domain.DeliveryAddress;
import br.com.ftgo.order.domain.Order;
import br.com.ftgo.order.domain.OrderItem;
import br.com.ftgo.order.domain.event.OrderEvents;

/**
 * Casos de uso do ServicoOrder (bounded context Order Management,
 * <em>core domain</em> do FTGO — ver EC-M2.1 e EC-M4.3).
 *
 * <p>API síncrona exposta (EC-M4.3): {@code POST /orders},
 * {@code GET /orders/{id}}. Eventos consumidos: CreditReserved,
 * CreditRejected, TicketPrepared, DeliveryScheduled (+ DeliveryCompleted,
 * extensão documentada em {@link Order}). Nesta primeira versão, esses
 * eventos chegam via um endpoint HTTP de simulação, no lugar de um
 * consumidor Kafka real, enquanto os demais microsserviços não
 * existem.</p>
 */
@Service
public class OrderService {

    private final OrderRepository repository;
    private final EventPublisher eventPublisher;

    public OrderService(OrderRepository repository, EventPublisher eventPublisher) {
        this.repository = repository;
        this.eventPublisher = eventPublisher;
    }

    public Order createOrder(String consumerId, String restaurantId,
                             List<OrderItem> items, DeliveryAddress deliveryAddress) {
        Order order = Order.create(consumerId, restaurantId, items, deliveryAddress);
        repository.save(order);
        eventPublisher.publish(order.toCreatedEvent());
        return order;
    }

    public Order getOrder(String id) {
        return repository.findById(id).orElseThrow(() -> new OrderNotFoundException(id));
    }

    public List<Order> listOrders() {
        return repository.findAll();
    }

    public Order handleCreditReserved(String orderId, long reservedAmountCents) {
        Order order = getOrder(orderId);
        OrderEvents.OrderApproved event = order.onCreditReserved(reservedAmountCents);
        repository.save(order);
        eventPublisher.publish(event);
        return order;
    }

    public Order handleCreditRejected(String orderId, String reason) {
        Order order = getOrder(orderId);
        OrderEvents.OrderCancelled event = order.onCreditRejected(reason);
        repository.save(order);
        eventPublisher.publish(event);
        return order;
    }

    public Order cancelOrder(String orderId, String reason) {
        Order order = getOrder(orderId);
        OrderEvents.OrderCancelled event = order.cancel(reason);
        repository.save(order);
        eventPublisher.publish(event);
        return order;
    }

    public Order handleTicketPrepared(String orderId) {
        Order order = getOrder(orderId);
        order.onTicketPrepared();
        repository.save(order);
        return order;
    }

    public Order handleDeliveryScheduled(String orderId, String courierId) {
        Order order = getOrder(orderId);
        order.onDeliveryScheduled(courierId);
        repository.save(order);
        return order;
    }

    public Order handleDeliveryCompleted(String orderId) {
        Order order = getOrder(orderId);
        OrderEvents.OrderDelivered event = order.onDeliveryCompleted();
        repository.save(order);
        eventPublisher.publish(event);
        return order;
    }
}
